feat: added database

dimaAndriyanov opened the pull request:
Добавлено использование реляционной базы данных h2

dimaAndriyanov committed:
feat: added database

dimaAndriyanov committed:
feat: added database

dimaAndriyanov committed:
Merge remote-tracking branch 'origin/add-database' into add-database

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java

dimaAndriyanov committed:
Merge remote-tracking branch 'origin/add-database' into add-database

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java

dimaAndriyanov committed:
Merge remote-tracking branch 'origin/add-database' into add-database

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java

Kapinosva commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,161 @@
+
+@Repository
+@Slf4j
+public class FilmDbStorage implements FilmStorage {
```
единообразно должно быть все. Если все DAO значит все DAO, если все storage, то все стораж

Kapinosva commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,161 @@
+@Repository
+@Slf4j
+public class FilmDbStorage implements FilmStorage {
+    private final JdbcTemplate jdbcTemplate;
```
это в каждом классе DB повоторяется - можно вынести в асбтрактный класс и везде его расширять

Kapinosva commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,161 @@
+                mpaRatingDao.getById(resultSet.getInt("mpa_rating_id"))
+        );
+        resultFilm.setId(resultSet.getInt("film_id"));
+        genreDao.getByFilmId(resultFilm.getId()).forEach(resultFilm::addGenre);
```
слишком много запросов к БД на каждый фильм, можем повесить БД

Kapinosva commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -0,0 +1,140 @@
+
+    @Override
+    public List<User> getAll() {
+        String sql = "select * from users";
```
всегда плохо *, лучше перечислить все поля

Kapinosva commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserMapper.java:
```diff
@@ -0,0 +1,38 @@
+        user.setName(resultSet.getString("user_name"));
+
+        String sql = "select friend_id from users_friends where user_id = ?";
+        jdbcTemplate.query(sql, (resSet, rowNum) -> resSet.getInt("friend_id"), user.getId())
```
каждый раз на каждом пользователе будет дополнительный запрос к БД. Лучше JOIN использовать

dimaAndriyanov committed:
style: renamed dao classes, corrected sql queries
fix: added abstract superclass for db storage
fix: reduced sql queries amount for get methods for users and films

dimaAndriyanov committed:
style: renamed dao classes, corrected sql queries
fix: added abstract superclass for db storage
fix: reduced sql queries amount for get methods for users and films

dimaAndriyanov committed:
style: renamed dao classes, corrected sql queries
fix: added abstract superclass for db storage
fix: reduced sql queries amount for get methods for users and films

Kapinosva commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,23 +1,24 @@
+        this.filmsLikesDbStorage = filmsLikesDbStorage;
     }
 
     public List<Film> getAll() {
```
сложновато, но работоспособно 


в идеальном варианте тут мы получаем список фильмов от FilmStorage - одни запросом все фильмы

Потом мы вызываем метод сервиса жанров которому передаем список фильмов.
Внутри сервиса жанров идет обращение к хранилищу жанров
Хранилище одним запросов вытаскивает все жанры всех переданных фмльмов (или просто все жанры всех фильмов)
далее сервис жанров устанавливает каждому фильму из переданного в метод списка фильмов нужный жанр.

и так по аналогии в каждом подобном примере.

для популярных аналогично, только хранилище фильмов нам отдаст не все фильмы, а только популярные - дальнейшие действия как выше.

Если нужны лайки, т отак же

Kapinosva commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,237 @@
+    public void deleteAllLikesFromUserById(int id) {
+    }
+
+    private List<Film> mapFilms(SqlRowSet rowSet) {
```
сложновато вышло

Kapinosva reviewed (commented):
сложновато, но работоспособно 


в идеальном варианте тут мы получаем список фильмов от FilmStorage - одни запросом все фильмы

Потом мы вызываем метод сервиса жанров которому передаем список фильмов.
Внутри сервиса жанров идет обращение к хранилищу жанров
Хранилище одним запросов вытаскивает все жанры всех переданных фмльмов (или просто все жанры всех фильмов)
далее сервис жанров устанавливает каждому фильму из переданного в метод списка фильмов нужный жанр.

и так по аналогии в каждом подобном примере.

для популярных аналогично, только хранилище фильмов нам отдаст не все фильмы, а только популярные - дальнейшие действия как выше.

Если нужны лайки, т отак же
